syntax = "proto3";
package DataProvider;

message CtrlPingRequest {
  string input = 1;
}

message CtrlPingResponse {
	string output = 1;
}

message TimeInterval {
  string initial_date = 1;
  string end_date = 2;
}

message TickerData {
  string ticker = 1;
  map<string, string> indicators_to_values = 2;
}

message DayData{
  string date = 1;
  repeated TickerData ticker_data = 2;
}

message StockData {
  repeated DayData multiple_days_data = 1;
}

message PastDataRequest {
  repeated string tickers = 1;
  repeated string indicators = 2;
  TimeInterval interval = 3;
}

message PastDataResponse {
  oneof Response {
    StockData data = 1;
    TimeInterval interval = 2;
  }
  enum Status {
    OK = 0;
    DATE_NOT_ALLOWED = 1;
    NOK = 2;
  }

  Status status = 3;
}

/*

message ExchangeDailyData{

}

message CurrentDailyDataRequest {
  ExchangeDailyData exchange_data = 1;
}

message CurrentDailyDataResponse {
  StockData daily_data = 1;
  enum Status {
    OK = 0;
    NOK = 1;
  }
} */

service DataProvider {
  rpc ctrl_ping(CtrlPingRequest) returns (CtrlPingResponse);
  rpc get_past_data(PastDataRequest) returns (stream PastDataResponse);
  // rpc get_daily_data(CurrentDailyDataRequest) returns (CurrentDailyDataResponse);
  // rpc get_intra_daily_data(CurrentIntraDailyDataRequest) returns (CurrentIntraDailyDataResponse);
}